internal/controller/pullrequest_controller.go

```diff
@@ -58,121 +58,92 @@ type PullRequestReconciler struct {
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/reconcile
 func (r *PullRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
+	logger.Info("Reconciling PullRequest")
 
 	var pr promoterv1alpha1.PullRequest
-	err := r.Get(ctx, req.NamespacedName, &pr, &client.GetOptions{})
-	if err != nil {
+	if err := r.Get(ctx, req.NamespacedName, &pr); err != nil {
 		if errors.IsNotFound(err) {
 			logger.Info("PullRequest not found", "namespace", req.Namespace, "name", req.Name)
 			return ctrl.Result{}, nil
 		}
-
-		logger.Error(err, "failed to get PullRequest", "namespace", req.Namespace, "name", req.Name)
 		return ctrl.Result{}, fmt.Errorf("failed to get PullRequest: %w", err)
 	}
 
-	pullRequestProvider, err := r.getPullRequestProvider(ctx, pr)
+	provider, err := r.getPullRequestProvider(ctx, pr)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed to get PullRequest provider: %w", err)
 	}
-	if pullRequestProvider == nil {
-		return ctrl.Result{}, fmt.Errorf("failed to get pull request provider, pullRequestProvider is nil")
-	}
 
-	deleted, err := r.handleFinalizer(ctx, &pr, pullRequestProvider)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to handle finalizer: %w", err)
-	}
-	if deleted {
-		return ctrl.Result{}, nil
+	if deleted, err := r.handleFinalizer(ctx, &pr, provider); err != nil || deleted {
+		return ctrl.Result{}, err
 	}
 
-	found, err := pullRequestProvider.FindOpen(ctx, &pr)
+	logger.Info("Checking for open PR on provider")
+	found, err := provider.FindOpen(ctx, &pr)
 	if err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed to check for open PR: %w", err)
 	}
 
-	// We can't find an open PR on the provider, and we have had a status state before, we should now delete it because
-	// it no longer exists on provider.
 	if !found && pr.Status.State != "" {
```

Comment:
We should probably add docs on the `State` field to clarify that it should only be populated based on a response from the SCM confirming the actual state. Or we could switch to something that less likely to be accidentally populated, like `ID`. Both are vanishingly unlikely to be misused. But probably worth clarifying to avoid subtle bugs.

We should probably also document that the PullRequest lifecycle is strictly tied to the SCM PullRequest. If the user closes or merges the PR, we delete the PullRequest resource. That documentation could probably just go on this package instead of in docs/.

Reply:
Zach pointed out we don't delete until a reconcile, and we don't really reconcile on a regular interval. Just good to document whatever our current choice is.